Compiling error when building without SLEPC

## Problem description
The eigenproblem class is heavily ifdefed on LIBMESH_HAVE_SLPEPC because it does not make much sense to solve eigenproblems without our eigensolver. But a virtual class is declared and never defined if that variable is false. Hence a compilation error

## How to reproduce
Compile without slepc

## Impact
User issue when compiling a complex project, seemingly without slepc

### Discussed in https://github.com/idaholab/moose/discussions/30453

<div type='discussions-op-text'>

<sup>Originally posted by **sblondel** May  5, 2025</sup>
### Check these boxes if you have followed the posting rules.

- [X] I have consulted the [Posting Guidelines](https://github.com/idaholab/moose/discussions/18270).
- [X] I have searched the [Discussions Forum](https://github.com/idaholab/moose/discussions) and [MOOSE Framework Troubleshooting](https://mooseframework.inl.gov/help/troubleshooting.html) and have not found what I was looking for
- [X] Q&A Getting Started is the most appropriate category for my question (trouble installing, beginner user, ...)

### Issue or question about MOOSE

Hi,

I am working on the HPC (Sawtooth) system and NOT using the provided MOOSE environment on purpose. My final goal is to have an external code [Xolotl](https://github.com/ORNL-Fusion/xolotl) that is already wrapped as a [MOOSEapp](https://github.com/SciDAC-MOOSE-Xolotl-coupling-group/coupling_xolotl) all compiled with the same dependencies, in particular PETSc.

The most recent version of Xolotl uses the Kokkos version of PETSc, so I am letting Xolotl build the PETSc version that it uses, then defining `PETSC_DIR` so that libmesh and MOOSE can use that PETSc. Everything goes well until the linking stage:
```
WARNING: Failed to initialize PreMake for version checking; this may be ignored but may suggest an environment issue.

Traceback (most recent call last):
  File "/projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/../scripts/premake.py", line 20, in __init__
    from versioner import Versioner
  File "/projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/scripts/versioner.py", line 26, in <module>
    from graphlib import TopologicalSorter
ModuleNotFoundError: No module named 'graphlib'

Traceback (most recent call last):
  File "/projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/../scripts/premake.py", line 258, in <module>
    PreMake().check()
  File "/projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/../scripts/premake.py", line 27, in __init__
    self.packages = Versioner().get_packages('HEAD')
UnboundLocalError: local variable 'Versioner' referenced before assignment
make: [/projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/moose.mk:455: prebuild] Error 1 (ignored)
Building and linking /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/capabilities/capabilities.so...
Compiling C++ (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/exodiff/STRINGLIB_tokenize.C...
Compiling C++ (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/exodiff/create_file.C...
Compiling C++ (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/exodiff/Tolerance.C...
Compiling C++ (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/exodiff/ED_SystemInterface.C...
...
Compiling C++ (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/build/unity_src/interfaces_Unity.C...
Compiling C++ (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/pugixml/src/pugixml.C...
Compiling C++ (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/minijson/src/minijson/parse.C...
Compiling C++ (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/minijson/src/minijson/minijson.C...
Compiling C++ (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/minijson/src/minijson/jsonValue.C...
Compiling C++ (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/tinyhttp/src/tinyhttp/http.C...
Compiling C (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/mtwist/src/randistrs.c...
Compiling C (in opt mode) /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/mtwist/src/mtwist.c...
Linking Library /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/contrib/hit/libhit-opt.la...
Linking Library /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/test/lib/libmoose_test-opt.la...
Linking Library /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/libmoose-opt.la...
Linking Executable /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/test/moose_test-opt...
/projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/libmoose-opt.so: undefined reference to `typeinfo for EigenProblem'
/projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/libmoose-opt.so: undefined reference to `vtable for EigenProblem'
collect2: error: ld returned 1 exit status
make: *** [/projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/framework/app.mk:466: /projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/moose/test/moose_test-opt] Error 1
```

and here are the modules I have loaded:
```
Currently Loaded Modules:
  1) gmp/6.1.2-gcc-10.5.0-wep3    4) zstd/1.4.4-gcc-10.5.0-yue3    7) curl/7.61.1-gcc-12.3.0-5t73  10) cmake/3.27.7-gcc-12.3.0-5cfk
  2) mpfr/4.2.0-gcc-10.5.0-cec4   5) gcc/12.3.0-gcc-10.5.0-vx2f    8) ncurses/6.1-gcc-12.3.0-7zps
  3) mpc/1.3.1-gcc-10.5.0-mpcb    6) mvapich2/2.3.7-1-gcc-12.3.0   9) zlib/1.2.13-gcc-12.3.0-c2ww
```

I would be happy to discuss other options to have both codes working together if you think my approach is not compatible with the current MOOSE. I could define a minimal Conda environment for instance.

Thank you for your help,

Sophie

### (Optional) code in question / simulation log / errors

_No response_

### Encountering Errors? Please include diagnostic output

```pre
####################################################################################################
Influential Environment Variables

CC=mpicc
CONDA_CHANNEL=https://conda.software.inl.gov/public
CURL_CA_BUNDLE=
CXX=mpicxx
F77=mpif77
F90=mpif90
FC=mpif90
LD_LIBRARY_PATH=/apps/local/mvapich2/2.3.7-gcc-12.3.0/lib
MODULESHOME=/apps/spack/opt/gcc-12.3.0/lmod-8.7.32-jpqjeet2bt3lyiiwrph7mo6iiznyuiea/lmod/lmod
MOOSE_JOBS=6
PATH=/apps/spack/opt/gcc-12.3.0/cmake-3.27.7-5cfktbfycm7h3l77524abhbkxakzpjyn/bin:/apps/local/mvapich2/2.3.7-gcc-12.3.0/bin:/apps/spack/opt/gcc-10.5.0/gcc-12.3.0-vx2ffftzrhvsqiflq3hwu2nissbw5dpf/bin:/opt/clmgr/sbin:/opt/clmgr/bin:/opt/sgi/sbin:/opt/sgi/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/c3/bin:/opt/pbs/bin
PETSC_DIR=/projects/NE_SciDAC/LDRD/coupling_xolotl_kokkos/xolotl/build/external/petsc_install
REQUESTS_CA_BUNDLE=
SSL_CERT_FILE=
####################################################################################################
Compiler(s) (CC CXX FC F77 F90):

which $CC; /apps/local/mvapich2/2.3.7-gcc-12.3.0/bin/mpicc
$CC --version; gcc (Spack GCC) 12.3.0
$CC -show; gcc -fPIC -g -O -fno-omit-frame-pointer -I/apps/local/mvapich2/2.3.7-gcc-12.3.0/include -L/apps/local/mvapich2/2.3.7-gcc-12.3.0/lib -Wl,-rpath -Wl,/apps/local/mvapich2/2.3.7-gcc-12.3.0/lib -Wl,--enable-new-dtags -lmpi

which $CXX; /apps/local/mvapich2/2.3.7-gcc-12.3.0/bin/mpicxx
$CXX --version; g++ (Spack GCC) 12.3.0
$CXX -show; g++ -g -O -fPIC -std=gnu++11 -I/apps/local/mvapich2/2.3.7-gcc-12.3.0/include -L/apps/local/mvapich2/2.3.7-gcc-12.3.0/lib -lmpicxx -Wl,-rpath -Wl,/apps/local/mvapich2/2.3.7-gcc-12.3.0/lib -Wl,--enable-new-dtags -lmpi

which $FC; /apps/local/mvapich2/2.3.7-gcc-12.3.0/bin/mpif90
$FC --version; GNU Fortran (Spack GCC) 12.3.0
$FC -show; gfortran -fPIC -g -O -fallow-argument-mismatch -I/apps/local/mvapich2/2.3.7-gcc-12.3.0/include -I/apps/local/mvapich2/2.3.7-gcc-12.3.0/include -L/apps/local/mvapich2/2.3.7-gcc-12.3.0/lib -lmpifort -Wl,-rpath -Wl,/apps/local/mvapich2/2.3.7-gcc-12.3.0/lib -Wl,--enable-new-dtags -lmpi

which $F77; /apps/local/mvapich2/2.3.7-gcc-12.3.0/bin/mpif77
$F77 --version; GNU Fortran (Spack GCC) 12.3.0
$F77 -show; gfortran -fPIC -g -O -fallow-argument-mismatch -I/apps/local/mvapich2/2.3.7-gcc-12.3.0/include -I/apps/local/mvapich2/2.3.7-gcc-12.3.0/include -L/apps/local/mvapich2/2.3.7-gcc-12.3.0/lib -lmpifort -Wl,-rpath -Wl,/apps/local/mvapich2/2.3.7-gcc-12.3.0/lib -Wl,--enable-new-dtags -lmpi

which $F90; /apps/local/mvapich2/2.3.7-gcc-12.3.0/bin/mpif90
$F90 --version; GNU Fortran (Spack GCC) 12.3.0
$F90 -show; gfortran -fPIC -g -O -fallow-argument-mismatch -I/apps/local/mvapich2/2.3.7-gcc-12.3.0/include -I/apps/local/mvapich2/2.3.7-gcc-12.3.0/include -L/apps/local/mvapich2/2.3.7-gcc-12.3.0/lib -lmpifort -Wl,-rpath -Wl,/apps/local/mvapich2/2.3.7-gcc-12.3.0/lib -Wl,--enable-new-dtags -lmpi

OK
####################################################################################################
Python Sanity Checks

/usr/bin/env python3 --version; (reporting as: Python 3.6.8) matches
which python3 python;

/usr/bin/python3 --version; == Python 3.6.8
/usr/bin/python --version; == Python 3.6.8

OK
####################################################################################################
Python Modules (TestHarness, run-ability)


FAIL:    missing module(s): packaging 
WARNING: missing module(s): jinja2 

Either install the above packages, or perhaps you have yet
to activate the moose environment: conda activate moose

Missing 'failing' Python modules will prevent you from building MOOSE.
Missing 'warning' Python modules may cause miscellaneous runtime issues.

FAIL: One or more Python issues present.
```
</div>

--- a/framework/src/problems/EigenProblem.C
+++ b/framework/src/problems/EigenProblem.C
@@ -695,10 +695,10 @@ EigenProblem::formNorm()
               "We should not get here unless a bx_norm postprocessor has been provided");
   return getPostprocessorValueByName(*_bx_norm_name);
 }
+#endif
 
 std::string
 EigenProblem::solverTypeString(const unsigned int solver_sys_num)
 {
   return Moose::stringify(solverParams(solver_sys_num)._eigen_solve_type);
 }
-#endif

